src/nvim/decoration.c

```diff
@@ -792,50 +789,95 @@ DecorSignHighlight *decor_find_sign(DecorInline decor)
   }
 }
 
-// Increase the signcolumn size and update the sentinel line if necessary for
-// the invalidated range.
-void decor_validate_signcols(buf_T *buf, int max)
+/// Add overlapping signs in current row to count. If it is greater than
+/// current max, set it and make b_signcols valid. Adjust max_count if
+/// row has or had max signs.
+void buf_signcols_count_row(buf_T *buf, const int overlap[], int idx, int count, bool del)
```

Comment:
`overlap` and `idx` are only used once. I would move the count adjustment outside to make this function less convoluted.